gnovm/cmd/gno/repl.go

```diff
@@ -107,68 +114,176 @@ func runRepl(cfg *replCfg) error {
 		handleInput(r, cfg.initialCommand)
 	}
 
-	fmt.Fprint(os.Stdout, "gno> ")
+	fmt.Fprint(os.Stdout, gnoREPL)
+
+	var (
+		inEdit      bool
+		prev        string
+		indentLevel int
+	)
 
-	inEdit := false
-	prev := ""
 	liner := bufio.NewScanner(os.Stdin)
 
 	for liner.Scan() {
 		line := liner.Text()
 
-		if l := strings.TrimSpace(line); l == ";" {
-			line, inEdit = "", false
-		} else if l == "/editor" {
-			line, inEdit = "", true
-			fmt.Fprintln(os.Stdout, "// enter a single ';' to quit and commit")
-		}
+		trimmedLine := strings.TrimSpace(line)
+
+		indentLevel = updateIndentLevel(trimmedLine, indentLevel)
+		line, inEdit = handleEditor(line)
+
 		if prev != "" {
 			line = prev + "\n" + line
 			prev = ""
 		}
+
 		if inEdit {
-			fmt.Fprint(os.Stdout, "...  ")
+			fmt.Fprint(os.Stdout, inEditMode)
 			prev = line
+
 			continue
 		}
 
 		if err := handleInput(r, line); err != nil {
 			var goScanError scanner.ErrorList
 			if errors.As(err, &goScanError) {
-				// We assune that a Go scanner error indicates an incomplete Go statement.
+				// We assume that a Go scanner error indicates an incomplete Go statement.
 				// Append next line and retry.
 				prev = line
 			} else {
 				fmt.Fprintln(os.Stderr, err)
 			}
 		}
 
-		if prev == "" {
-			fmt.Fprint(os.Stdout, "gno> ")
-		} else {
-			fmt.Fprint(os.Stdout, "...  ")
-		}
+		printPrompt(indentLevel, prev)
 	}
+
 	return nil
 }
 
+func handleEditor(line string) (string, bool) {
+	if l := strings.TrimSpace(line); l == ";" {
+		return "", false
+	} else if l == editorCommand {
+		fmt.Fprintln(os.Stdout, "// enter a single ';' to quit and commit")
+
+		return "", true
+	}
+
+	return line, false
+}
+
+func updateIndentLevel(line string, indentLevel int) int {
+	openCount, closeCount := 0, 0
+	increaseIndent := false
+
+	for i, char := range line {
+		switch char {
+		case '{', '(', '[':
+			openCount++
+
+			if i < len(line)-1 && line[i+1] == '\n' {
+				increaseIndent = true
+			}
+		case '}', ')', ']':
+			closeCount++
+		}
+	}
+
+	indentLevel += openCount - closeCount
+	if indentLevel < 0 {
+		indentLevel = 0
+	}
+
+	if increaseIndent {
+		indentLevel++
+	}
+
+	if strings.HasSuffix(line, ":") {
```

Comment:
Maybe I'm being too pedantic here but this can cause some odd behavior as well:
```go
gno> a := A{a:
...         4
...         b:
...             5
...             }
...         println(a)
...         
```

Reply:
Oh, I didn't see that, good catch 👍

Reply:
Fixed so that indentation does not change if a colon follows the end of a string. If possible, I'd like to make it so that only the body part of the `case`, `default` are indented, but I'd probably have to switch the terminal to raw mode to do that.

```plain
gno> func main() {
...      num := 1
...      switch num {
...          case 1:
...          println("one")
...          default:
...          println("not one")
...          }
...      }
```